compiler/semantic/op.go

```diff
@@ -596,10 +612,33 @@ func (a *analyzer) semOp(o ast.Op, seq dag.Seq) (dag.Seq, error) {
 			Limit: o.Limit,
 		}), nil
 	case *ast.Put:
-		assignments, err := a.semAssignments(o.Args, false)
+		assignments, err := a.semAssignments(o.Args)
 		if err != nil {
 			return nil, err
 		}
+		// We can do collision checking on static paths, so check what we can.
+		var fields field.List
+		for _, a := range assignments {
+			if this, ok := a.LHS.(*dag.This); ok {
+				fields = append(fields, this.Path)
+			}
+		}
+		for i, f := range fields {
+			if f.IsEmpty() {
+				return nil, fmt.Errorf("put: LHS cannot be 'this' (use 'yield' operator)")
+			}
+			for j, c := range fields {
+				if i == j {
+					continue
+				}
+				if f.Equal(c) {
+					return nil, fmt.Errorf("put: multiple assignments to %s", f)
+				}
+				if c.HasStrictPrefix(f) {
+					return nil, fmt.Errorf("put: conflicting nested assignments to %s and %s", f, c)
+				}
+			}
+		}
```

Comment:
This is a nearly verbatim copy of runtime/expr/checkFields. Export that as CheckPutFields or similar and call it here.

Reply:
Good idea